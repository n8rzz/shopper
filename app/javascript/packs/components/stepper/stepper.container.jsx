import React from 'react';
import PropTypes from 'prop-types';
import { Stepper } from './stepper';

export class StepperContainer extends React.Component {
    constructor(props) {
        super(props);

        this.onClickDecreaseHandler = this._onClickDecrease.bind(this);
        this.onClickIncreaseHandler = this._onClickIncrease.bind(this);
        this.onClickSubmitHandler = this._onClickSubmit.bind(this);

        this.state = {
            qty: 1,
        };
    }

    componentWillUnmount() {
        this.onClickDecreaseHandler = null;
        this.onClickIncreaseHandler = null;
        this.onClickSubmitHandler = null;
    }

    _onClickDecrease() {
        // eslint-disable-next-line react/no-access-state-in-setstate
        let nextQty = this.state.qty - 1;

        if (nextQty < 1) {
            nextQty = 1;
        }

        this.setState({ qty: nextQty });
    }

    _onClickIncrease() {
        this.setState((prevState) => ({ qty: prevState.qty + 1 }));
    }

    _onClickSubmit(event) {
        event.stopPropagation();
        event.preventDefault();

        if (this.state.qty < 1) {
            return;
        }

        this.props.onSubmitHandler(this.state.qty);

        this.setState({ qty: 1 });
    }

    render() {
        return (
            <Stepper
                itemId={this.props.itemId}
                qty={this.state.qty}
                isSubmitting={this.props.isSubmitting}
                isSubmitSuccess={this.props.isSubmitSuccess}
                onClickDecreaseHandler={this.onClickDecreaseHandler}
                onClickIncreaseHandler={this.onClickIncreaseHandler}
                onClickSubmitHandler={this.onClickSubmitHandler}
            />
        );
    }
}

StepperContainer.propTypes = {
    itemId: PropTypes.number.isRequired,
    isSubmitting: PropTypes.bool.isRequired,
    isSubmitSuccess: PropTypes.bool.isRequired,
    onSubmitHandler: PropTypes.func.isRequired,
};
