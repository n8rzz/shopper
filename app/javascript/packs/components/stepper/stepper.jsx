import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

export class Stepper extends React.Component {
    buildClassnames() {
        return classnames({
            'stepper-ft-submit': true,
            'mix-btn_isLoading': this.props.isSubmitting,
            'mix-btn_isSuccess': this.props.isSubmitSuccess,
        });
    }

    _buildStepperBody() {
        return (
            <div className={'stepper-bd'}>
                <div
                    className={'stepper-bd-btn js-stepper-btn-decrease'}
                    onClick={this.props.onClickDecreaseHandler}
                >
                    -
                </div>
                <div id={`item-${this.props.itemId}-qty`}>
                    {this.props.qty}
                </div>
                <div
                    className={'stepper-bd-btn js-stepper-btn-increase'}
                    onClick={this.props.onClickIncreaseHandler}
                >
                    +
                </div>
            </div>
        );
    }

    _buildSubmitButton() {
        let buttonValue = this.props.isSubmitting
            ? 'Adding...'
            : 'Add to order';

        if (this.props.isSubmitSuccess) {
            buttonValue = 'Success!';
        }

        return (
            <div className={'stepper-ft'}>
                <input
                    type={'submit'}
                    value={buttonValue}
                    className={this.buildClassnames()}
                />
            </div>
        );
    }

    render() {
        return (
            <form onSubmit={this.props.onClickSubmitHandler}>
                <div
                    className={'stepper'}
                    data-stepper-id={this.props.itemId}
                >
                    { this._buildStepperBody() }
                    { this._buildSubmitButton() }
                </div>
            </form>
        );
    }
}

Stepper.propTypes = {
    itemId: PropTypes.number.isRequired,
    qty: PropTypes.number.isRequired,
    isSubmitting: PropTypes.bool.isRequired,
    isSubmitSuccess: PropTypes.bool.isRequired,
    onClickDecreaseHandler: PropTypes.func.isRequired,
    onClickIncreaseHandler: PropTypes.func.isRequired,
    onClickSubmitHandler: PropTypes.func.isRequired,
};
