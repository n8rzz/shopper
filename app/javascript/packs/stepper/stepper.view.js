const CLASSNAMES = {
    ROOT: 'js-stepper',
    DECREASE: 'js-stepper-decrease',
    VALUE: 'js-stepper-value',
    INCREASE: 'js-stepper-increase',
    SUBMIT: 'js-stepper-submit',
};

const ACTION = {
    DECREASE: 'DECREASE',
    INCREASE: 'INCREASE',
};

/**
 * Connects to a number input with external decrease/increase
 * action elements in addition to a submit button
 *
 * accepts `onSubmitHandler` from parent
 *
 * This class is concerned with decreasin/increasing a value of
 * a number field, and submitting that current value onClick
 *
 * @class StepperView
 */
export default class StepperView {
    constructor($element, onSubmitHandler) {
        this._id = -1;
        this._$element = $element;
        this._onSubmitHandler = onSubmitHandler;

        if (!this._$element) {
            return null;
        }

        return this._createChildren()
            ._setupHandlers()
            .enable();
    }

    _createChildren() {
        this._id = parseInt(this._$element.dataset.stepperId, 10);
        this._$decreaseButton = document.getElementsByClassName(this._buildClassnameWithId(CLASSNAMES.DECREASE))[0];
        this._$increaseButton = document.getElementsByClassName(this._buildClassnameWithId(CLASSNAMES.INCREASE))[0];
        this._$valueView = document.getElementsByClassName(this._buildClassnameWithId(CLASSNAMES.VALUE))[0];
        this._$submitButton = document.getElementsByClassName(this._buildClassnameWithId(CLASSNAMES.SUBMIT))[0];

        return this;
    }

    _setupHandlers() {
        this._onClickDecreaseHandler = this._onClickDecrease.bind(this);
        this._onClickIncreaseHandler = this._onClickIncrease.bind(this);
        this._onClickSubmitHandler = this._onClickSubmit.bind(this);

        return this;
    }

    enable() {
        this._$decreaseButton.addEventListener('click', this._onClickDecreaseHandler);
        this._$increaseButton.addEventListener('click', this._onClickIncreaseHandler);
        this._$submitButton.addEventListener('click', this._onClickSubmitHandler);

        return this;
    }

    _buildClassnameWithId(classname) {
        return `${classname}-${this._id}`;
    }

    _onClickDecrease(event) {
        this._updateValueView(ACTION.DECREASE);
    }

    _onClickIncrease(event) {
        this._updateValueView(ACTION.INCREASE);
    }

    _onClickSubmit(event) {
        const currentValue = parseInt(this._$valueView.value, 10);

        this._onSubmitHandler(event, currentValue);
        this._resetValue();
    }

    _resetValue() {
        this._$valueView.value = 1;
    }

    _updateValueView(action) {
        const currentValue = parseInt(this._$valueView.value, 10);
        let nextValue;

        switch(action) {
            case ACTION.DECREASE:
                if (currentValue <= 1) {
                    return;
                }

                nextValue = currentValue - 1;

                break;
            case ACTION.INCREASE:
                nextValue = currentValue + 1;

                break;
            default:
                break;
        }

        this._$valueView.value = nextValue;
    }
}
