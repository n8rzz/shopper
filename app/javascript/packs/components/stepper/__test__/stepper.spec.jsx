import React from 'react';
import { shallow } from 'enzyme';
import { Stepper } from '../stepper';

describe('Stepper', () => {
    let component;

    beforeEach(() => {
        component = shallow(<Stepper
            itemId={42}
            qty={3}
            isSubmitting={false}
            isSubmitSuccess={false}
            onChangeQtyHandler={jest.fn()}
            onClickDecreaseHandler={jest.fn()}
            onClickIncreaseHandler={jest.fn()}
            onClickSubmitHandler={jest.fn()}
        />);
    });

    afterEach(() => {
        component = null;
    });

    test('renders correctly with valid props', () => {
        expect(component).toMatchSnapshot();
    });

    test('contains decrease btn element', () => {
        const decreaseBtn = component.find('.js-stepper-btn-decrease');

        expect(decreaseBtn.length).toBe(1);
    });

    test('contains increase btn element', () => {
        const increaseBtn = component.find('.js-stepper-btn-increase');

        expect(increaseBtn.length).toBe(1);
    });

    test('contains add to order btn element', () => {
        const submitBtn = component.find('.stepper-ft-submit');

        expect(submitBtn.length).toBe(1);
    });

    describe('when #isSubmitting is true', () => {
        let isSubmittingComponent;

        beforeEach(() => {
            isSubmittingComponent = shallow(<Stepper
                itemId={42}
                qty={3}
                isSubmitting={true}
                isSubmitSuccess={false}
                onChangeQtyHandler={jest.fn()}
                onClickDecreaseHandler={jest.fn()}
                onClickIncreaseHandler={jest.fn()}
                onClickSubmitHandler={jest.fn()}
            />);
        });

        afterEach(() => {
            isSubmittingComponent = null;
        });

        test('it should add correct className to btn', () => {
            const submitBtn = isSubmittingComponent.find('.mix-btn_isLoading');

            expect(submitBtn.length).toBe(1);
        });

        test('it should render correct btn value', () => {
            const submitBtn = isSubmittingComponent.find('.stepper-ft-submit');

            expect(submitBtn.props().value).toBe('Adding...');
        });
    });

    describe('when #isSubmitSuccess is true', () => {
        let isSubmitSucessComponent;

        beforeEach(() => {
            isSubmitSucessComponent = shallow(<Stepper
                itemId={42}
                qty={3}
                isSubmitting={false}
                isSubmitSuccess={true}
                onChangeQtyHandler={jest.fn()}
                onClickDecreaseHandler={jest.fn()}
                onClickIncreaseHandler={jest.fn()}
                onClickSubmitHandler={jest.fn()}
            />);
        });

        afterEach(() => {
            isSubmitSucessComponent = null;
        });

        test('it should add correct className to btn', () => {
            const submitBtn = isSubmitSucessComponent.find('.mix-btn_isSuccess');

            expect(submitBtn.length).toBe(1);
        });

        test('it should render correct btn value', () => {
            const submitBtn = isSubmitSucessComponent.find('.stepper-ft-submit');

            expect(submitBtn.props().value).toBe('Success!');
        });
    });
});
