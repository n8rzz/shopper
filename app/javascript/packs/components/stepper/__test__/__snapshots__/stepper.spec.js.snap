// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stepper renders correctly with valid props 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <div
    className="stepper"
    data-stepper-id={42}
  >
    <div
      className="stepper-bd"
    >
      <label
        className="stepper-bd-label u-isVisuallyHidden"
        htmlFor="qty"
      >
        Qty
      </label>
      <span
        className="stepper-bd-btn js-stepper-btn-decrease"
        onClick={[MockFunction]}
      >
        -
      </span>
      <input
        className="stepper-bd-input"
        id="item-42-qty"
        min="0"
        name="item-42-qty"
        onChange={[MockFunction]}
        pattern="[0-9]*"
        type="number"
        value={3}
      />
      <span
        className="stepper-bd-btn js-stepper-btn-increase"
        onClick={[MockFunction]}
      >
        +
      </span>
    </div>
    <div
      className="stepper-ft"
    >
      <input
        className="stepper-ft-submit"
        type="submit"
        value="Add to order"
      />
    </div>
  </div>
</form>
`;
